### PR TITLE
[SPARK-16856] [WEBUI] [CORE] Link the application's executor page to the master's UI

### PR DESCRIPTION
**Pull request by nblintao**
## What changes were proposed in this pull request?

Added a "**Back to Master**" link on Executor Page of Application Detail UI, if the application is run in standalone mode.

The links link to the main page of Master UI.
## How was this patch tested?

Run Spark as a "local cluster" by `./bin/spark-shell --master=local-cluster[2,1,1024]`.

Then view the Executor Page on Application Detail UI (http://192.168.1.104:4040/executors/). Users could go to the Master Page by the link "**Back to Master**" on that page.

detailpagetomaster

### CODE REVIEW

**Comment by ajbozarth** [comment]
I take some time to check out and run this later today, but on a quick pass the code looks ok. I have some issues with the actual pr though. The link on the application summary page is redundant, clicking the spark logo goes back to the master summary already. As for the Executors page, I'll have to test it, but I think there are quite a few situations where a admin might not want a user to access the master UI, even in a standalone cluster.

**Comment by ajbozarth** [comment]
I forgot to comment, but I did check this out and it runs well, I still think the link on the app summary is redundant though.

**Comment by nblintao** [comment]
@ajbozarth Thanks for pointing out that. I have removed the link on the application summary page and reverted related changes.

**Comment by nblintao** [comment]
retest it please

**Comment by nblintao** [comment]
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/project/MimaExcludes.scala:50: ')' expected but '.' found.
[error]       ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.SparkListenerApplicationStart.apply")
[error]                     ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/project/MimaExcludes.scala:51: ';' expected but '.' found.
[error]       ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.scheduler.SparkListenerApplicationStart.copy")
[error]                     ^
```
I think I have fixed the conflict correctly. Why's that?

**Comment by ajbozarth** [comment]
The excludes are a list, you forgot to add commas between them

**Comment by nblintao** [comment]
Oh yes. Thanks :)

**Comment by ajbozarth** [comment]
@nblintao I'd like to take another look at this, mind rebasing?

**Comment by nblintao** [comment]
@ajbozarth I finally have a chance to rebase it in the winter break. Could you please have a look? Thanks!

**Comment by ajbozarth** [comment]
Thanks for rebasing, I'll check this out sometime this week

**Comment by ajbozarth** [comment]
@nblintao did you try this code on different cluster types? Like local, standalone, yarn, mesos?

**Comment by ajbozarth** [comment]
Also @srowen and @vanzin could you take a look at this one too?

**Comment by ajbozarth** [comment]
I did some testing and this properly shows up on local and standalone and doesn't show the link on yarn. So this LGTM

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala [hunk 1]
```diff
@@ -126,7 +126,8 @@ case class SparkListenerApplicationStart(
     time: Long,
     sparkUser: String,
     appAttemptId: Option[String],
-    driverLogs: Option[Map[String, String]] = None) extends SparkListenerEvent
+    driverLogs: Option[Map[String, String]] = None,
+    masterWebUiUrl: Option[String] = None) extends SparkListenerEvent
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: Do you need this? It seems the standalone registration messages already take care of this.

If you want this, are you intentionally not saving this in event logs? The way it is, this is only ever useful when the application is starting but hasn't yet registered with the master. That's a very short period of time and probably not worth it.

If you add code to save to event logs, I'd rename this to be more cluster manager-agnostic. e.g. "clusterWebUiUrl".

File: core/src/main/scala/org/apache/spark/ui/exec/ExecutorsPage.scala [hunk 2]
```diff
@@ -50,6 +50,10 @@ private[ui] class ExecutorsPage(
   private val listener = parent.listener
 
   def render(request: HttpServletRequest): Seq[Node] = {
+    val masterWebUiUrl = listener.masterWebUiUrl
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/exec/ExecutorsPage.scala]: I think this information makes way more sense in the "Environment" tab.

**Comment by HyukjinKwon** [comment]
(gentle ping @nblintao)

**Comment by nblintao** [comment]
@HyukjinKwon Thanks for reminding. I will have a look asap. Too many things to recently.

**Comment by HyukjinKwon** [comment]
(gentle ping @nblintao)

**Comment by nblintao** [comment]
@HyukjinKwon Thanks for reminding. I am finals this week. I will fix this after 15th.

**Pull request closed without merging**